### PR TITLE
esi: Restore the original onerror behavior

### LINKED ISSUE #4053: Parent ESI fails even when VCL sets resp.status = 200 for the included ESI
### Expected Behavior

[What's New](https://varnish-cache.org/docs/trunk/whats-new/changes-7.3.html#esi-processing-changes) for Varnish 7.3 says of ESI includes:

> ... only objects with 200 and 204 status will be included and any other status code will fail the parent ESI request.
> 
> If objects with other status should be delivered, they should have their status changed to 200 in VCL, for instance in sub vcl_backend_error{}, vcl_synth{} or vcl_deliver{}.

So we expected that this snippet in `vcl_deliver{}` would prevent failure of the including parent in all cases:

```
sub vcl_deliver {
    if (req.esi_level > 0 && resp.status != 200 && resp.status != 204) {
        set resp.status = 200;
    }
} 
```

### Current Behavior

Even with the VCL snippet shown above, the parent ESI can evidently fail under certain circumstances, in fact apparently in default configuration. I'll add a VTC test case, based on Varnish's current `e00035.vtc`, to illustrate that.

The VTC suggests that `beresp.do_stream` may be related to the problem. If `beresp.do_stream` is set to false and the VCL snippet is included in `vcl_deliver`, then as expected, the parent response does not fail, and the client response body is not truncated. If the `beresp.do_stream == true`, which is the default, then the parent response fails and the body is truncated, even with the `vcl_deliver` snippet.

That would at least make sense to me, since it's familiar that when streaming is on, a client response status may be seen as 200, even when the fetch actually failed (for example when there were too few bytes for the value in `Content-Length`), and would have been set to 503 if streaming had been off. On that logic, `vcl_deliver` doesn't "see" `resp.status != 200` when that happens.

But the VTC shows a bit of strangeness, because it also seems to depend on whether or not `beresp.do_esi` is conditionally or unconditionally set to `true` in VCL. The problem happens if `beresp.do_esi` is set conditionally, as in `e00035.vtc`. If VCL just says `set beresp.do_esi = true`, then the problem doesn't happen in the VTC.

Obviously that takes some explanation. I'll go through the expectations when I add the VTC test case.

### Possible Solution

In `vcl_backend_response`, set `beresp.do_esi = true` unconditionally. It may also be necessary to set `beresp.do_stream = false` (but in that case we lose the advantages of streaming).

### Steps to Reproduce (for bugs)

I'll add a VTC test case to the issue, and describe what we expected to see.

### Context

The site makes very extensive use of ESI, and has been prepared to handle errors in ESI subrequests for quite some time before the attempted migration to Varnish 7.x.

We know of `onerror="continue"` and `-p feature=+esi_include_onerror`. The Varnish team can control the `-p` parameter, but using the `onerror` attribute in responses depends on all of the developer teams (about 50 of them) always remembering to do so, without exception for every ESI include, now and in the future. One omission and the entire response fails. So the VCL snippet in `vcl_deliver` is intended to safeguard against that.

### Varnish Cache version

varnishd (varnish-7.4.2 revision cd1d10ab53a6f6115b2b4f3b2a1da94c1f749f80)

### Operating system

Debian bookworm (12.4)

### Source of binary packages used (if any)

Both packagecloud and build from github source

### PR DESCRIPTION
**Pull request by Dridi**
This is a partial revert of 582ded6a2d6ae1a4467b1eb500f2725b42888016 to restore the assumed onerror=continue behavior for ESI includes, unless the feature flag esi_include_onerror is raised.

The part of the change that considers all status codes besides 200 and 204 to be errors for ESI includes remains. A test case covers VCL's ability to "bless" error responses by overriding resp.status, allowing ESI delivery to continue on this criterion.

Fixes #4053

### CODE REVIEW

**Comment by nigoroll** [comment]
Note: if we merge this, we should also have #4054

**Comment by nigoroll** [comment]
> The part of the change that considers all status codes besides 200 and 204 to be errors for ESI includes remains.

But also only if `esi_include_onerror` is enabled, right?

**Comment by Dridi** [comment]
> > The part of the change that considers all status codes besides 200 and 204 to be errors for ESI includes remains.
> 
> But also only if `esi_include_onerror` is enabled, right?

Correct, it adds (or leaves alone) one more error condition eligible to abort. But the feature flag must be raised to stop defaulting to "continue", just like when said feature flag was introduced.

**Comment by Dridi** [comment]
> I think the decision to be made here is if we want more explicit control as suggested in [#4053 (comment)](https://github.com/varnishcache/varnish-cache/issues/4053#issuecomment-1941693695) (tri-state `esi_include_onerror = attribute | abort | continue`) or if we are happy with just "taming" the feature as suggested here.
> 
> My personal opinion is that this PR _together with #4054_ is the minimum we should do, but Varnish-Cache has its strength in the flexibility of its configuration and I think that it might very well be possible to get to a point were we want VCL to override to `abort`.

My opinion is that this PR is the minimum we should do, that is, restoring the previous behavior while retaining the "new" status code effect.

If we want more than a flag, then we should consider promoting it to its own parameter. Regarding other ESI feature flags, we could also consider a new `feature_esi` parameter, similar to what we did with `feature_vcc` (except that in this case we couldn't leave compatibility aliases behind).

**Comment by nigoroll** [comment]
Interesting additional complication uncovered by @xcir [here](https://gitlab.com/uplex/varnish/libvdp-pesi/-/issues/16#note_1795415262): What happens when an included object is neither a 200 nor a 204, but has an empty body?

**Comment by slimhazard** [comment]
@nigoroll isn't that entirely covered by the existing requirements?

I'd say: if the `esi_include_onerror` feature is off, then the top-level response is aborted. If it's on, and the esi tag has `onerror="continue"` or `resp.status` is set to 200 or 204, then include the empty response body. Since it's empty, it looks like nothing in the top-level rendered response body, which is presumably the same thing that happens for an included response with status 204.

All of which is to say "do what Varnish does now" (or is intended to do, assuming a fix for #4053). How is it a complication?

**Comment by Dridi** [comment]
> Interesting additional complication uncovered by @xcir [here](https://gitlab.com/uplex/varnish/libvdp-pesi/-/issues/16#note_1795415262): What is supposed to happen when an included object is neither a 200 nor a 204, but has an empty body?

For the sake of the argument, I'll say the same behavior since Varnish 3.0, which is my oldest reference.

The default behavior is an implicit onerror="continue" attribute, so the erroneous status code of an ESI fragment does not influence the req_top delivery, whether the ESI fragment has a body or not.

**Comment by nigoroll** [comment]
> if the `esi_include_onerror` feature is off, then the top-level response is aborted.

I tend to agree, but this is not [what we currently do](https://github.com/varnishcache/varnish-cache/blob/master/bin/varnishd/cache/cache_esi_deliver.c#L871-L872).

**Comment by slimhazard** [comment]
@nigoroll hmm ... I've worked with ESIs that return status 204 and an empty body, since Varnish 3.0.x (same site as the one I've been talking about). Those have always rendered 0 bytes into the including response. An empty included response does something different now, if status != 20x? That would be very strange (and again I'm puzzled as to why).

**Comment by nigoroll** [comment]
extra-bugwash: merge

File: bin/varnishtest/tests/r03241.vtc [hunk 1]
```diff
@@ -43,11 +47,10 @@ logexpect l1 -v v1 -g raw {
 
 client c1 {
 	txreq -hdr "Host: foo"
-	rxresphdrs
+	rxresp
+	# XXX this is actually wrong (missed include)
+	expect resp.bodylen == 57
```
Comment (Dridi) [inline bin/varnishtest/tests/r03241.vtc:52]: This comment should be changed:

> 15:24 < phk> so this comment should really say "if you want this to be correct, set esi_include_onerror ?

It will be done in a separate commit to keep the revert minimal.

**Review by Dridi — COMMENTED** [review]
bugwash: merge

**Comment by nigoroll** [comment]
brought up by phk: we need to adjust docs

**Comment by Dridi** [comment]
Rebased against master to give cci a headstart while working on docs.

**Comment by Dridi** [comment]
Rebased against current master, most notably to incorporate #4071.

**Review by nigoroll — APPROVED** [review]
thank you!
